bolt-modules/boltlib/lib/puppet/datatypes/resourceinstance.rb

```diff
@@ -16,7 +16,8 @@
       overwrite_state         => Callable[[Hash[String[1], Data]], Hash[String[1], Data]],
       set_desired_state       => Callable[[Hash[String[1], Data]], Hash[String[1], Data]],
       overwrite_desired_state => Callable[[Hash[String[1], Data]], Hash[String[1], Data]],
-      reference               => Callable[[], String]
+      reference               => Callable[[], String],
+      '[]'                    => Callable[[String[1]], Optional[Data]]
```

Comment:
`Data` is an alias for a `Variant` that includes `Undef`, so this doesn't need to be `Optional`.